plugins/out_s3/s3.c

```diff
@@ -980,28 +1045,39 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
     flb_sds_destroy(s3_key);
     uri = tmp;
 
+    if (ctx->use_put_object == FLB_TRUE && ctx->compression != NULL) {
+        if (strcmp(ctx->compression, "gzip") != 0) {
+            flb_plg_error(ctx->ins, 
+                          "'gzip' is the only support value of compression field for now");
+            flb_sds_destroy(uri);
+            return -1;
+        }
+        ret = flb_gzip_compress(body, body_size, &final_body, &final_body_size);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "Failed to compress data");
+            flb_sds_destroy(uri);
+            return -1;
+        }
+    } else {
+        final_body = body;
+        final_body_size = body_size;
+    }
+    
     s3_client = ctx->s3_client;
     if (s3_plugin_under_test() == FLB_TRUE) {
         c = mock_s3_call("TEST_PUT_OBJECT_ERROR", "PutObject");
     }
     else {
-        if (ctx->canned_acl == NULL) {
-            c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                                  uri, body, body_size,
-                                                  NULL, 0);
-        }
-        else {
-            canned_acl_header = create_canned_acl_header(ctx->canned_acl);
-            if (canned_acl_header == NULL) {
-                flb_sds_destroy(uri);
-                flb_plg_error(ctx->ins, "Failed to create canned ACL header");
-                return -1;
-            }
-            c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                                  uri, body, body_size,
-                                                  canned_acl_header, 1);
-            flb_free(canned_acl_header);
+        ret = create_headers(ctx, &headers, &num_headers);
+        if (ret == -1 || (headers == NULL && num_headers != 0)) {
```

Comment:
the check should just be `if (ret < 0)`

your `create_headers` function should return `-1` for all failure cases